aws-crt-http/HttpConnection.cs

```diff
@@ -1,8 +1,245 @@
+        }
+
+        internal class Handle : CRT.Handle
+        {
+            protected override bool ReleaseHandle()
+            {
+                API.destroy(handle);
+                return true;
+            }
+        }
+
+        internal Handle NativeHandle { get; private set; }
+
+        public HttpClientConnection Connection { get; protected set; }
+        public int ResponseStatusCode { get; protected set; }
+
+        // reference to options used to create this stream, which keeps the callbacks alive
+        // for the duration of the stream
+        private HttpRequestOptions options;
+
+        public HttpStream(HttpClientConnection connection, HttpRequestOptions options) 
+        {
+            if (options.OnIncomingHeaders == null)
+                throw new ArgumentNullException("OnIncomingHeaders");
+            if (options.OnStreamComplete == null)
+                throw new ArgumentNullException("OnStreamComplete");
+
+            this.Connection = connection;
+
+            // Wrap the native callbacks to bind this stream to them as the first argument
+            OnStreamOutgoingBodyNative onStreamOutgoingBody = (out byte[] buffer, out UInt64 size) =>
+            {
+                options.OnStreamOutgoingBody(this, out buffer);
+                size = (UInt32)buffer.Length;
+            };
+            OnIncomingHeadersNative onIncomingHeaders = (headers, headerCount) =>
+            {
+                options.OnIncomingHeaders(this, headers);
+            };
+            OnIncomingHeaderBlockDoneNative onIncomingHeaderBlockDone = (hasBody) =>
+            {
+                options.OnIncomingHeaderBlockDone(this, hasBody);
+            };
+            OnIncomingBodyNative onIncomingBody = (data, size) =>
+            {
+                options.OnIncomingBody(this, data);
+            };
+            OnStreamCompleteNative onStreamComplete = (errorCode) =>
+            {
+                options.OnStreamComplete(this, errorCode);
+            };
+            NativeHandle = API.make_new(
+                connection.NativeHandle.DangerousGetHandle(),
+                options.Method,
+                options.Uri.PathAndQuery,
+                options.Headers,
+                (UInt32)(options.Headers?.Length ?? 0),
+                options.OnStreamOutgoingBody != null ? onStreamOutgoingBody : null,
+                onIncomingHeaders,
+                options.OnIncomingHeaderBlockDone != null ? onIncomingHeaderBlockDone : null,
+                options.OnIncomingBody != null ? onIncomingBody : null,
+                onStreamComplete);
+            this.options = options;
+        }
+
+        public void UpdateWindow(uint incrementSize)
+        {
+            API.update_window(NativeHandle.DangerousGetHandle(), incrementSize);
+        }
+    }
+
+    public sealed class HttpClientStream : HttpStream
+    {
+        public HttpClientStream(HttpClientConnection connection, HttpRequestOptions options)

```

Comment:
Would it be a better user experience if request creation was a method on the Connection object? I think that's what henso did in aws-crt-cpp. something like:

HttpClientConnection::SendRequest(HttpRequestOptions) -> HttpClientStream

Reply:
I can do both!